File: app/src/main/java/top/maweihao/weather/WeatherApplication.java
package top.maweihao.weather;


import android.text.TextUtils;

import com.tencent.bugly.crashreport.CrashReport;

import org.litepal.LitePalApplication;

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;

import top.maweihao.weather.util.Constants;
import top.maweihao.weather.util.LogUtils;
import top.maweihao.weather.util.Utility;

/**
 * Created by limuyang on 2017/12/13.
 */

public class WeatherApplication extends LitePalApplication {

    private static final String TAG = WeatherApplication.class.getSimpleName();

    @Override
    public void onCreate() {
        super.onCreate();

        String pn = getProcessName(android.os.Process.myPid());
        if (pn == null || pn.equals(getPackageName())) {
            initBugly();
            initAdmob();
            initTimeAndLang();
            initLog();
        }
    }

    private void initLog() {
        LogUtils.init(this);
        LogUtils.getConfig().setLogSwitch(BuildConfig.APP_DEBUG);
    }

    private void initTimeAndLang() {
        Constants.timeShift = Utility.getTimeShift();
        Constants.lang = Utility.getCurrentLanguage(this);
//        Log.d("Application", "time shift and lang is " + Constants.timeShift + Constants.lang);
    }

    private void initBugly() {
        CrashReport.initCrashReport(getApplicationContext(), "2af8412ed0", BuildConfig.APP_DEBUG, getStrategy());
    }

    private void initAdmob() {
//        MobileAds.initialize(this, Constants.AD_MOB_ID);
    }

    private static String getProcessName(int pid) {
        BufferedReader reader = null;
        try {
            reader = new BufferedReader(new FileReader("/proc/" + pid + "/cmdline"));
            String processName = reader.readLine();
            if (!TextUtils.isEmpty(processName)) {
                processName = processName.trim();
            }
            return processName;
        } catch (Throwable throwable) {
            throwable.printStackTrace();
        } finally {
            try {
                if (reader != null) {
                    reader.close();
                }
            } catch (IOException exception) {
                exception.printStackTrace();
            }
        }
        return null;
    }

    private CrashReport.UserStrategy getStrategy() {
        CrashReport.UserStrategy strategy = new CrashReport.UserStrategy(getApplicationContext());
        // TODO: 2018/5/27  
        return strategy;
    }
}

File: app/src/main/java/top/maweihao/weather/WeatherApplication.kt
package top.maweihao.weather


import android.text.TextUtils
import com.blankj.utilcode.util.LogUtils
import com.blankj.utilcode.util.Utils
import com.tencent.bugly.crashreport.CrashReport
import org.litepal.LitePalApplication
import top.maweihao.weather.util.Constants
import top.maweihao.weather.util.Utility
import java.io.BufferedReader
import java.io.FileReader
import java.io.IOException

/**
 * Created by limuyang on 2017/12/13.
 */

class WeatherApplication : LitePalApplication() {
    override fun onCreate() {
        super.onCreate()

        val pn = getProcessName(android.os.Process.myPid())
        if (pn == null || pn == packageName) {
            initBugly()
            initAdmob()
            initTimeAndLang()
            initUtils()
        }
    }

    private fun initUtils() {
        Utils.init(this)
        LogUtils.getConfig().setLogSwitch(BuildConfig.APP_DEBUG)
    }

    private fun initTimeAndLang() {
        Constants.timeShift = Utility.getTimeShift()
        Constants.lang = Utility.getCurrentLanguage(this)
        //        Log.d("Application", "time shift and lang is " + Constants.timeShift + Constants.lang);
    }

    private fun initBugly() {
        CrashReport.initCrashReport(applicationContext, "2af8412ed0", BuildConfig.APP_DEBUG, getStrategy())
    }

    private fun initAdmob() {
        //        MobileAds.initialize(this, Constants.AD_MOB_ID);
    }

    private fun getProcessName(pid: Int): String? {
        var reader: BufferedReader? = null
        try {
            reader = BufferedReader(FileReader("/proc/$pid/cmdline"))
            var processName = reader.readLine()
            if (!TextUtils.isEmpty(processName)) {
                processName = processName.trim { it <= ' ' }
            }
            return processName
        } catch (throwable: Throwable) {
            throwable.printStackTrace()
        } finally {
            try {
                if (reader != null) {
                    reader.close()
                }
            } catch (exception: IOException) {
                exception.printStackTrace()
            }

        }
        return null
    }

    private fun getStrategy(): CrashReport.UserStrategy {
        // TODO: 2018/5/27
        return CrashReport.UserStrategy(applicationContext)
    }
}
